### ISSUE TITLE
Output Cache returns cached result when it shouldn't - Cache Control: Public

### ISSUE THREAD

**Issue by dcinzona** [OrchardCMS/Orchard#6587]
We are able to reproduce this consistently:
1. In the Cache Settings, set the `max-age` property value to something  > 0 (we did 3600)
2. Open a new private browsing window and navigate to your site, hitting a page and creating a cached-copy
3. Log into the site
4. Navigate back to the page that you had hit before with the cached copy

What happens is you get the browser cached version of the page, instead of hitting the server to get the server cached version, effectively causing you to be logged out

I believe this is because of the Cache-Control header option: https://github.com/OrchardCMS/Orchard/blob/dev/src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs#L505

`response.Cache.SetCacheability(HttpCacheability.Public);` will store the cached copy in the browser, and if the browser has a cached copy, then it never sends the request to the server, and so the user remains logged out until they hit refresh, effectively logging them back in.

I'm not sure if this is by design, or if changing `.Public` to `.Server` would resolve the issue, but it might.

`.Server` Summary
```
// Summary:
//     Specifies that the response is cached only at the origin server. Similar to the
//     System.Web.HttpCacheability.NoCache option. Clients receive a Cache-Control:
//     no-cache directive but the document is cached on the origin server. Equivalent
//     to System.Web.HttpCacheability.ServerAndNoCache.
```

**Comment by dcinzona** [comment]
Actually, it may actually be that this line is what breaks auth with regards to output caching:
`response.Cache.SetOmitVaryStar(true);`

Confirmed in our QA environment.  Removing that line resolves the issue.

**Comment by jtkech** [comment]
@dcinzona

I don't remember the details, but as i remember, indeed, when public caching (if maxAge > 0) it was to be able to use client browser caching, not only proxy servers, CDNs or IIS kernel cache.

And that's why, on a post redirection, if maxAge > 0, a refresh key is used in the rediretct url. But, indeed, if you use the same url of a page cached on client, you will get the same content. Note that it could be the same before with another public cache as a proxy...

In the dashboard, see the comments related to this setting. As @DanielStolt suggested, maybe we would need a maxAge per individual routes. Otherwise, you would need to set maxAge to 0.

`Specifies the duration (in seconds) content is allowed to be cached on the client, and on intermediaries such as proxy servers, CDNs or IIS kernel cache.`
`When set to a non-zero value, a cache-control header with a max-age property will be added to responses.`
`NOTE: Responses cached on intermediaries or clients will not be invalidated when content is edited, and cannot participate in the full cache key variation supported in Orchard.`
`Only activate this if you really know what you're doing, and consider setting it to a considerably lower value than the default cache duration.`


Best.

**Comment by dcinzona** [comment]
So how would this not break authenticated site browsing?  
Take this example:
1. Go to site anonymously
2. Go to About-Us page (or some other page on the site)
3. Go to another page via main navigation
4. Log in
5. Go back to About-Us page by clicking on the main navigation link on the page you are currently on

The About-Us page is pulled from the local cache, where you are not authenticated, thus returning an un-authenticated view.

**Comment by sebastienros** [comment]
Yes, max-age and SetOmitVaryBy will break authenticated cached pages. So if your cached content can vary when authenticated, then you should not use the SetOmitVaryBy. It's not an issue for pages that are not cached in the first place, like the admin. It's an issue only if you render authenticated content in your front end.

**Comment by dcinzona** [comment]
So could there be another checkbox / setting for this option, making it a declaration, with it's own set of hints, instead of adding it to the default max-age option?

**Comment by jtkech** [comment]
Yes, i understand. But does it works as before if you set maxAge to 0?

**Comment by dcinzona** [comment]
Yes, it does.  But aren't you bundling two different caching options here into one setting?

**Comment by jtkech** [comment]
Not sure, i have to take a look. But as i remember it's only to set public cache and use a refresh key on port redirection.

**Comment by dcinzona** [comment]
I don't follow the `on port redirection` part.  In regards to the PR, that was to fix a declarative header vary-by option.  Instead of making it declarative, i.e. have the user fill in a list of headers to vary cache responses by, the Omit(*) option ignores all, does it not?  I think it needs additional review.

**Comment by jtkech** [comment]
On post redirection, a refresh key is added to the url when maxAge > 0, this not to use a public cache.

But, i agree, it's a "breaking" change for those that was used to set maxAge > 0. But it could happen before with another public cache (more rarely). And some browsers (not sure with modern ones) may not disable caching even with a vary star.

But again, not sure, i've no close opinion, just try to understand. Maybe we need an option for the vary star. Or remove this line, but it was the only way to be able to use client caching.

Best

**Comment by sebastienros** [comment]
We definitely need an option for SetVaryBy, the question is do max-age without SetVaryBy make sense. 
1- Yes, then options are distinct for max-age and SetVaryBy
2- No, then a single option

Then alternate question: do we want this/these options global or per route.

**Comment by dcinzona** [comment]
If we are going to propose changing the per-route options, I would add an additional per-route option: `Ignore Authentication State` ... i.e. the route should **always** return the cached result if it exists, regardless if the user is authenticated or not.

**Comment by dcinzona** [comment]
Or to keep the same nomenclature: `vary by authentication state` with the default = true
